tests/cases/compiler/deferredCallbacks.ts

```diff
@@ -0,0 +1,238 @@
+// @strict: true
+// @declaration: true
+// @target: esnext
+
+declare function immediate(cb: () => void): void;
+declare function deferred1(deferred cb: () => void): void;
+declare function deferred2(/** @deferred */ cb: () => void): void;
+declare function deferred3(/** @deferred */ deferred cb: () => void): void;
+
+function f01() {
+    let x: string | number = "OK";
+    immediate(() => {
+        x = 42;
+    });
+    x;  // string | number
+}
+
+function f02() {
+    let x: string | number = "OK";
+    deferred1(() => {
+        x = 42;
+    });
+    x;  // string
+}
+
+function f03() {
+    let x: string | number = "OK";
+    deferred2(() => {
+        x = 42;
+    });
+    x;  // string
+}
+
+function f04() {
+    let x: string | number = "OK";
+    deferred3(() => {
+        x = 42;
+    });
+    x;  // string
+}
+
+// Parameter is considered deferred if one or more overloads defer that parameter
+
+declare function overloaded<T>(cb: (x: T) => T): void;
+declare function overloaded<T>(cb: (x: T, y: T) => T): void;
+declare function overloaded(deferred cb: (...args: any) => any): void;
+
+function f05() {
+    let x: string | number = "OK";
+    overloaded(() => {
+        x = 42;
+    });
+    x.length;
+}
+
+// deferred is permitted on a rest parameter
+
+declare function invokeImmediate(...args: ((...args: any) => any)[]): void;
+declare function invokeDeferred(deferred ...args: ((...args: any) => any)[]): void;
+
+function f06() {
+    let a = [];
+    a.push("abc");
+    a;  // string[]
+    invokeImmediate(
+        () => {
+            a;  // string[]
+            a.push(42);
+            a;  // (string | number)[]
+        },
+        () => {
+            a;  // string[]
+            a.push(true);
+            a;  // (string | boolean)[]
+        }
+    );
+    a;  // (string | number | boolean)[]
+}
+
+function f07() {
+    let a = [];
+    a.push("abc");
+    a;  // string[]
+    invokeDeferred(
+        () => {
+            a;  // string[]
+            a.push(42);
+            a;  // (string | number)[]
+        },
+        () => {
+            a;  // string[]
+            a.push(true);
+            a;  // (string | boolean)[]
+        }
+    );
+    a;  // string[]
+}
+
+// deferred modifier must precede public/private/protected/readonly
+
+class CC {
+    constructor(deferred public readonly x: () => void) {}

```

Comment:
Just to make sure, it would be good to have tests of a construct *call*. I didn't see any but I might have missed them.

Reply:
Yeah; testing a variety of types with annotated construct signatures would be good too, like

```ts
interface Ctor1 {
  new(deferred cb: () => void): {}
}

type Ctor2 = new (deferred cb: () => void) => void;
```